www/react-base/src/components/BuildSummaryTooltip/BuildSummaryTooltip.tsx

```diff
@@ -0,0 +1,245 @@
+/*
+  This file is part of Buildbot.  Buildbot is free software: you can
+  redistribute it and/or modify it under the terms of the GNU General Public
+  License as published by the Free Software Foundation, version 2.
+
+  This program is distributed in the hope that it will be useful, but WITHOUT
+  ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or FITNESS
+  FOR A PARTICULAR PURPOSE.  See the GNU General Public License for more
+  details.
+
+  You should have received a copy of the GNU General Public License along with
+  this program; if not, write to the Free Software Foundation, Inc., 51
+  Franklin Street, Fifth Floor, Boston, MA 02110-1301 USA.
+
+  Copyright Buildbot Team Members
+*/
+
+import {useContext, useMemo, useState} from "react";
+import {ConfigContext} from "../../contexts/Config";
+import {observer} from "mobx-react";
+import {Step} from "../../data/classes/Step";
+import {useDataAccessor, useDataApiQuery} from "../../data/ReactUtils";
+import {Builder} from "../../data/classes/Builder";
+import {Build} from "../../data/classes/Build";
+import {results2class, results2text} from "../../util/Results";
+import {durationFormat, useCurrentTime} from "../../util/Moment";
+
+const isStepDisplayed = (step: Step) => {
+  return !step.hidden;
+}
+
+const limitStringLength = (s: string, limit: number) => {
+  let res = s.slice(0, limit);
+  if (s.length > limit) {
+    res += ' ...';
+  }
+  return res;
+}
+
+type BuildInfoByUrls = {
+  builderid: string;
+  buildnumber: string;
+}
+
+type BuildRequestInfoByUrls = {
+  buildrequestid: string;
+}
+
+type StepInfoByUrls = {
+  buildrequests: BuildRequestInfoByUrls[];
+  builds: BuildInfoByUrls[];
+}
+
+const execRegexesGetFirstResult = (regexes: RegExp[], text: string) => {
+  for (const regex of regexes) {
+    const res = regex.exec(text)
+    if (res) {
+      return res
+    }
+  }
+  return null
+}
+
+type StepUrlAnalyzer = {
+  buildrequest: RegExp[];
+  build: RegExp[];
+};
+
+const useStepUrlAnalyzer = (baseUrls: string[]): StepUrlAnalyzer => {
+  const buildrequestRegexes = useMemo(
+    () => baseUrls.map(url => new RegExp(`${url}#buildrequests/([0-9]+)$`)),
+    [baseUrls]);
+  const buildRegexes = useMemo(
+    () => baseUrls.map(url => new RegExp(`${url}#builders/([0-9]+)/builds/([0-9]+)$`)),
+    [baseUrls]);
+
+  return {
+    buildrequest: buildrequestRegexes,
+    build: buildRegexes,
+  };
+}
+
```

Comment:
This file looks a bit complicated. I think it deserves some factorization

Reply:
Agreed. Currently everything is a direct rewrite from AngularJS side. I first tried to make it work and I will clean everything later.